sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java

```diff
@@ -0,0 +1,168 @@
+
+  @Nullable abstract String host();
+  abstract int port();
+  @Nullable abstract String auth();
+  abstract int timeout();
+
+  abstract Builder builder();
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setHost(String host);
+    abstract Builder setPort(int port);
+    abstract Builder setAuth(String auth);
+    abstract Builder setTimeout(int timeout);
+    abstract RedisConnectionConfiguration build();
+  }
+
+  public static RedisConnectionConfiguration create() {
+    return new AutoValue_RedisConnectionConfiguration.Builder()
+        .setHost(Protocol.DEFAULT_HOST)
+        .setPort(Protocol.DEFAULT_PORT)
+        .setTimeout(Protocol.DEFAULT_TIMEOUT).build();
+  }
+
+  public RedisConnectionConfiguration withHost(String host) {
+    checkArgument(host != null, "RedisConnectionConfiguration.create().withHost(host) called "
+        + "with empty host");
+    return builder().setHost(host).build();
+  }
+
+  public RedisConnectionConfiguration withPort(int port) {
+    checkArgument(port > 0, "RedisConnectionConfiguration.create().withPort(port) called with "
+        + "invalid port number (" + port + ")");
+    return builder().setPort(port).build();
+  }
+
+  public RedisConnectionConfiguration withAuth(String auth) {
+    checkArgument(auth != null, "RedisConnectionConfiguration.create().withAuth(auth) called "
+        + "with null auth");
+    return builder().setAuth(auth).build();
+  }
+
+  public RedisConnectionConfiguration withTimeout(int timeout) {
+    return builder().setTimeout(timeout).build();
+  }
+
+  /**
+   * Connect to the Redis instance.
+   */
+  public Jedis connect() {
+    return new Jedis(host(), port(), timeout());
+  }
+
+  protected boolean isClusterEnabled() {
+    try (Jedis connection = connect()) {
+      String[] infos = connection.info().split("\n");
+      for (String info : infos) {
+        if (info.startsWith("cluster_enabled:")) {
+          if (info.substring("cluster_enabled:".length()).contains("1")) {
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+  }
+
+  /**
+   * Populate the display data with connectionConfiguration details.
+   */
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder.addIfNotNull(DisplayData.item("host", host()));
+    builder.addIfNotNull(DisplayData.item("port", port()));
+    builder.addIfNotNull(DisplayData.item("timeout", timeout()));
+  }
+
+  public List<RedisService.RedisNode> getNodes() throws Exception {
+    if (this.isClusterEnabled()) {
+      return getClusterNodes();
+    }
+    return Collections.emptyList();
+  }
+
+  public List<RedisService.RedisNode> getClusterNodes() {
+    try (Jedis jedis = connect()) {
+      List<Object> slots = jedis.clusterSlots();
+
+      for (Object slotInfoObj : slots) {
+        List<Object> slotInfo = (List<Object>) slotInfoObj;
+
+        int startSlot = Integer.parseInt(slotInfo.get(0).toString());
```

Comment:
What type is slotInfo.get(0) actually? Seems like it represents an integer, but is a different type - is it a `String`? In that case a type cast to `String` may be more explicit